code/datums/components/laser_components.dm

```diff
@@ -0,0 +1,28 @@
+/datum/component/extralasers
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	var/path
+
+/datum/component/extralasers/Initialize(ammo, _lens, overlay)
+	if(!istype(parent, /obj/item/gun/energy/laser))
+		return COMPONENT_INCOMPATIBLE
+	var/obj/item/gun/energy/laser/L = parent
+	L.ammo_type  += new ammo (src)
+
+	qdel(_lens)
+	path = text2path( "/obj/item/external_lens/[overlay]")
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/detach)
+
+/datum/component/extralasers/proc/detach(datum/source, obj/item/I, mob/user, params)
```

Comment:
You need to make component inheritance call on this

Reply:
what